api/agent/agent.go

```diff
@@ -683,21 +685,60 @@ func (a *agent) runHot(ctx context.Context, call *call, tok ResourceToken, state
 			state.UpdateState(ctx, ContainerStateIdle, call.slots)
 			s := call.slots.queueSlot(&hotSlot{done, errC, container, nil})
 
-			select {
-			case <-s.trigger:
-			case <-time.After(time.Duration(call.IdleTimeout) * time.Second):
-				if call.slots.ejectSlot(s) {
-					logger.Info("Canceling inactive hot function")
-					return
-				}
-			case <-ctx.Done(): // container shutdown
-				if call.slots.ejectSlot(s) {
-					return
+			isFrozen := false
+			elapsed := time.Duration(0)
+			ticker := freezeTicker
+
+			for {
+				select {
+				case <-s.trigger: // slot already consumed
+				case <-ctx.Done(): // container shutdown
+				case <-a.shutdown: // server shutdown
+				case <-time.After(ticker):
+					elapsed += ticker
+					if elapsed < idleTimeout {
+
+						if ticker == freezeTicker {
+							if !isFrozen {
+								err := a.driver.Freeze(ctx, container)
+								if err != nil {
+									logger.WithError(err).Error("freeze error")
+									return
+								}
+								isFrozen = true
+							}
+							ticker = ejectTicker
+							continue
+						}
+
+						// if someone is waiting for resource in our slot queue, we must not terminate,
+						// otherwise, see if other slot queues have resource waiters.
+						stats := call.slots.getStats()
+						if stats.containerStates[ContainerStateWait] > 0 ||
+							a.resources.GetResourceTokenWaiterCount() <= 0 {
+							continue
+						}
+
+						logger.Debug("attempting hot function eject")
```

Comment:
maybe I need to eat something today, but this appears to attempt to eject the container after one ejectTicker interval after freezing rather than after the idle timeout? if so, why do we have idle timeout if we're just going to obey ejectTicker? should we set ejectTicker to idleTimeout or am I missing something here?

Reply:
sometimes i forget it's better to read pr comments afterwards. i see, we're ejecting only if nobody is hanging out. part of me thinks this defeats the purpose of idle to a certain degree. I do think we need to eject in certain cases (highlighted by the red TODO at the head of agent.go, among others) but this seems really aggressive. if i make 1 total api call every 2 seconds i'll have to eat the container start up time both times even though idle timeout is 30s, right?

Reply:
We are ejecting only:

1) if our own slot queue does not have any container-spawn code waiting on resources.
(eg. burst of route A calls should not trigger killing of route A containers.)
and
2) some other slot queue has someone blocked waiting on cpu & mem
(eg. burst of route B calls may kill route A containers, but only if route B blocks on cpu & memory condition variable and route A container was idling for 1 sec)

So, 1 total api call every 2 secs would only terminate that container if the timer fired and found out that there was some other route-call that was blocked on cpu & resources. If there is enough cpu & memory in the system, no call should block on the condition variable. They'd lock the mutex and get the resource... But if there's contention, then the call would likely to block on cond.wait().

Reply:
I take it that's the `a.resources.GetResourceTokenWaiterCount() <= 0`, blanked on that. that seems pretty slick to me.